### PR TITLE
feat: Add oauth client creation API & Mix Tasks

### LINKED ISSUE #375: [Task] Add the Lenra OAuth Client in a migration
There is an Hydra API endpoint to create the Lenra OAuth client.
Creating the "client" in the back-end is mandatory for the Lenra Client or any other third client to use the OAuth Endpoint.

See [Hydra API Doc](https://www.ory.sh/docs/hydra/reference/api#tag/oAuth2/operation/createOAuth2Client)

This PR must ensure that the Client will be created correctly before starting the Identity endpoint. This "migration" must be idempotent.
This is not a database migration but more like a mandatory idempotent HTTP request.

### PR DESCRIPTION
**Pull request by Nesqwik**
Closes #375

Also take care of future needs for the API.

## Description of the changes
Add a mix command to create the "backoffice" and "apps" oauth clients.
`mix create_oauth2_client <backoffice|apps|custom>`
- The "backoffice" or "apps" command should create the correct backoffice setup for dev or prod depending of the MIX_ENV value.
 - The "custom" command create a new oauth client without any default value
 - For any of the 3 commands, you can set/override the values

Example for custom with set values : 
```bash
mix create_oauth2_client custom   --name "My Custom Client"   --scope foo --scope bar   --redirect-uri http://example.com/redirect.html   --redirect-uri com.example.app:/oauth2redirect   --allowed-origin http://example.com   --allowed-origin http://auth.example.com
```

You also have the API routes to create the oauth2 clients via the Backoffice.
```
POST /api/environments/:environment_id/oauth2
GET /api/environments/:environment_id/oauth2
PUT /api/environments/:environment_id/oauth2/:client_id
DELETE /api/environments/:environment_id/oauth2/:client_id
```
 
## Checklist
- [ ] I didn't over-scope my PR (Well... A bit...)
- [x] My PR title matches the [commit convention](https://www.conventionalcommits.org/en/v1.0.0/)
- [x] I did not include breaking changes
- [x] I made my own code-review before requesting one

### I included unit tests that cover my changes
  - [x] 👍 yes
  - [ ] 🙅 no, because they aren't needed
  - [ ] 🙋 no, because I need help

### I added/updated the documentation about my changes

- [ ] 📜 README.md
- [x] 📕 docs/*.md
- [ ] 📓 docs.lenra.io
- [ ] 🙅 no documentation needed

## Technical highlight/advice
Nothing major. I've created a new table in the database that contains the environment_id, oauth2_client_id and the id of the association.
With that, you cas get all the oauth clients for a specific env_id.
Notes : 
- The `oauth2_client_id` is the ory hydra client_id.
- The `id` is basically not used. Just here for indexing.
- The couple `environment_id` `oauth2_client_id` is unique
- The Oauth client created with the mix command are not included in this table.

### CODE REVIEW

**Comment by taorepoara** [comment]
You should update the oauth doc to use this task. Also, the apps and backoffice clients creation is needed for the target use of the server. It should documented in the README.md two.

File: apps/lenra_web/lib/lenra_web/controllers/oauth2_controller.ex [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    with :ok <- allow_creator_only(conn, params),
+         {:ok, deleted} <- Apps.delete_oauth2_client(params) do
+      conn
+      |> reply(deleted |> Map.take([:oauth2_client_id, :environment_id]))
+    end
+  end
+
+  defp allow_creator_only(conn, %{"environment_id" => env_id}) do
+    with {:ok, app} <- Apps.fetch_app_for_env(env_id) do
+      allow(conn, app)
```
[the remark this answers is not in the document]
Comment (Nesqwik) [reply apps/lenra_web/lib/lenra_web/controllers/oauth2_controller.ex:45]: The `allow/2` function will call the `authorize/3` in the Policy below.
If the `authorize/3` return true, the authorization is granted (returns :ok). else, the authoziation is denied (return :error)

In this policy, the default `use LenraWeb.Policy.Default` will set everything that does not match before to false.
- The first argument is always the action (:index, :create, :update, :delete...).
- The second argument is always the current user.
- The third argument is whatever is passed to the `allow/2` second argument.

In this line in the Policy : 
```elixir
def authorize(_action, %User{id: user_id}, %App{creator_id: user_id}), do: true
```
For the function to match, the `user_id` had to be the same between the `%User{id: user_id}` and the `%App{creator_id: user_id}`. In other words, it return true if the current user_id is the same that the app creator_id.

**Comment by Nesqwik** [comment]
@taorepoara you should not merge alpha/beta to the current branch, only rebase the branch over alpha/beta :)

**Review by taorepoara — APPROVED** [review]
Works fine

**Pull request merged**